### ISSUE TITLE
Compilation seems to try to open a display

### ISSUE THREAD

**Issue by geokon-gh** [cljfx/cljfx#70]
this is a weird one...
I can build an uber jar locally in a terminal emulator, but if I try to do it through a terminal directly it won't build.

I discovered this by accident through Github Actions: 
https://github.com/geokon-gh/sausage/runs/640515413?check_suite_focus=true

I can replicate it locally as well by switching out of X (with `Ctrl`+`Alt`+`F2`) and trying to build an uberjar there (my setup is very similar to your setup in `hn`). The error reads

```
Caused by: java.lang.UnsupportedOperationException: Unable to open DISPLAY
```

And it's triggered by a namespace that seems to just define some effects functions. I also have a weird issues where when the application first launches it doesn't render immediately. I have to resize the window or move it around for the things to render. I feel it may be related

Anyways, if you've come across things - please let me know. Otherwise I'll keep trying to diagnose the issue and post a solution here for others to see once/if I do :)

**Comment by geokon-gh** [comment]
By running through Emacs/CIDER I managed to get the full error:

<details>
  <summary>The full error is</summary>

```
  Show: Project-Only All 
  Hide: Clojure Java REPL Tooling Duplicates  (36 frames hidden)

2. Unhandled clojure.lang.Compiler$CompilerException
   Error compiling cljfx/api.clj at (70:1)
   #:clojure.error{:phase :compile-syntax-check,
                   :line 70,
                   :column 1,
                   :source "cljfx/api.clj"}
             Compiler.java: 7647  clojure.lang.Compiler/load
                   RT.java:  381  clojure.lang.RT/loadResourceScript
                   RT.java:  372  clojure.lang.RT/loadResourceScript
                   RT.java:  463  clojure.lang.RT/load
                   RT.java:  428  clojure.lang.RT/load
                  core.clj: 6126  clojure.core/load/fn
                  core.clj: 6125  clojure.core/load
                  core.clj: 6109  clojure.core/load
               RestFn.java:  408  clojure.lang.RestFn/invoke
                  core.clj: 5908  clojure.core/load-one
                  core.clj: 5903  clojure.core/load-one
                  core.clj: 5948  clojure.core/load-lib/fn
                  core.clj: 5947  clojure.core/load-lib
                  core.clj: 5928  clojure.core/load-lib
               RestFn.java:  142  clojure.lang.RestFn/applyTo
                  core.clj:  667  clojure.core/apply
                  core.clj: 5985  clojure.core/load-libs
                  core.clj: 5969  clojure.core/load-libs
               RestFn.java:  137  clojure.lang.RestFn/applyTo
                  core.clj:  667  clojure.core/apply
                  core.clj: 6007  clojure.core/require
                  core.clj: 6007  clojure.core/require
               RestFn.java:  421  clojure.lang.RestFn/invoke
                 state.clj:    1  sausage.state/eval22019/loading--auto--
                 state.clj:    1  sausage.state/eval22019
                 state.clj:    1  sausage.state/eval22019
             Compiler.java: 7176  clojure.lang.Compiler/eval
             Compiler.java: 7165  clojure.lang.Compiler/eval
             Compiler.java: 7635  clojure.lang.Compiler/load
                   RT.java:  381  clojure.lang.RT/loadResourceScript
                   RT.java:  372  clojure.lang.RT/loadResourceScript
                   RT.java:  463  clojure.lang.RT/load
                   RT.java:  428  clojure.lang.RT/load
                  core.clj: 6126  clojure.core/load/fn
                  core.clj: 6125  clojure.core/load
                  core.clj: 6109  clojure.core/load
               RestFn.java:  408  clojure.lang.RestFn/invoke
                  core.clj: 5908  clojure.core/load-one
                  core.clj: 5903  clojure.core/load-one
                  core.clj: 5948  clojure.core/load-lib/fn
                  core.clj: 5947  clojure.core/load-lib
                  core.clj: 5928  clojure.core/load-lib
               RestFn.java:  142  clojure.lang.RestFn/applyTo
                  core.clj:  667  clojure.core/apply
                  core.clj: 5985  clojure.core/load-libs
                  core.clj: 5969  clojure.core/load-libs
               RestFn.java:  137  clojure.lang.RestFn/applyTo
                  core.clj:  667  clojure.core/apply
                  core.clj: 6007  clojure.core/require
                  core.clj: 6007  clojure.core/require
               RestFn.java:  421  clojure.lang.RestFn/invoke
               optical.clj:    1  sausage.optical/eval22013/loading--auto--
               optical.clj:    1  sausage.optical/eval22013
               optical.clj:    1  sausage.optical/eval22013
             Compiler.java: 7176  clojure.lang.Compiler/eval
             Compiler.java: 7165  clojure.lang.Compiler/eval
             Compiler.java: 7635  clojure.lang.Compiler/load
                   RT.java:  381  clojure.lang.RT/loadResourceScript
                   RT.java:  372  clojure.lang.RT/loadResourceScript
                   RT.java:  463  clojure.lang.RT/load
                   RT.java:  428  clojure.lang.RT/load
                  core.clj: 6126  clojure.core/load/fn
                  core.clj: 6125  clojure.core/load
                  core.clj: 6109  clojure.core/load
               RestFn.java:  408  clojure.lang.RestFn/invoke
                  core.clj: 5908  clojure.core/load-one
                  core.clj: 5903  clojure.core/load-one
                  core.clj: 5948  clojure.core/load-lib/fn
                  core.clj: 5947  clojure.core/load-lib
                  core.clj: 5928  clojure.core/load-lib
               RestFn.java:  142  clojure.lang.RestFn/applyTo
                  core.clj:  667  clojure.core/apply
                  core.clj: 5985  clojure.core/load-libs
                  core.clj: 5969  clojure.core/load-libs
               RestFn.java:  137  clojure.lang.RestFn/applyTo
                  core.clj:  667  clojure.core/apply
                  core.clj: 6007  clojure.core/require
                  core.clj: 6007  clojure.core/require
               RestFn.java:  703  clojure.lang.RestFn/invoke
                  core.clj:    1  sausage.core/eval7533/loading--auto--
                  core.clj:    1  sausage.core/eval7533
                  core.clj:    1  sausage.core/eval7533
             Compiler.java: 7176  clojure.lang.Compiler/eval
             Compiler.java: 7165  clojure.lang.Compiler/eval
             Compiler.java: 7635  clojure.lang.Compiler/load
                      REPL:    1  user/eval7529
                      REPL:    1  user/eval7529
             Compiler.java: 7176  clojure.lang.Compiler/eval
             Compiler.java: 7131  clojure.lang.Compiler/eval
                  core.clj: 3214  clojure.core/eval
                  core.clj: 3210  clojure.core/eval
                  main.clj:  414  clojure.main/repl/read-eval-print/fn
                  main.clj:  414  clojure.main/repl/read-eval-print
                  main.clj:  435  clojure.main/repl/fn
                  main.clj:  435  clojure.main/repl
                  main.clj:  345  clojure.main/repl
               RestFn.java: 1523  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   79  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:   55  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:  142  nrepl.middleware.interruptible-eval/interruptible-eval/fn/fn
                  AFn.java:   22  clojure.lang.AFn/run
               session.clj:  171  nrepl.middleware.session/session-exec/main-loop/fn
               session.clj:  170  nrepl.middleware.session/session-exec/main-loop
                  AFn.java:   22  clojure.lang.AFn/run
               Thread.java:  832  java.lang.Thread/run

1. Caused by java.lang.UnsupportedOperationException
   Unable to open DISPLAY

       GtkApplication.java:  173  com.sun.glass.ui.gtk.GtkApplication/lambda$new$6
     AccessController.java:  312  java.security.AccessController/doPrivileged
       GtkApplication.java:  171  com.sun.glass.ui.gtk.GtkApplication/<init>
   GtkPlatformFactory.java:   41  com.sun.glass.ui.gtk.GtkPlatformFactory/createApplication
          Application.java:  144  com.sun.glass.ui.Application/run
       QuantumToolkit.java:  280  com.sun.javafx.tk.quantum.QuantumToolkit/startup
         PlatformImpl.java:  269  com.sun.javafx.application.PlatformImpl/startup
             Platform.java:  103  javafx.application.Platform/startup
              platform.clj:    6  cljfx.jdk.platform/initialize
              platform.clj:    4  cljfx.jdk.platform/initialize
              platform.clj:   30  cljfx.platform/initialize
              platform.clj:   29  cljfx.platform/initialize
                   api.clj:   75  cljfx.api/eval24166
                   api.clj:   70  cljfx.api/eval24166
             Compiler.java: 7176  clojure.lang.Compiler/eval
             Compiler.java: 7635  clojure.lang.Compiler/load
                   RT.java:  381  clojure.lang.RT/loadResourceScript
                   RT.java:  372  clojure.lang.RT/loadResourceScript
                   RT.java:  463  clojure.lang.RT/load
                   RT.java:  428  clojure.lang.RT/load
                  core.clj: 6126  clojure.core/load/fn
                  core.clj: 6125  clojure.core/load
                  core.clj: 6109  clojure.core/load
               RestFn.java:  408  clojure.lang.RestFn/invoke
                  core.clj: 5908  clojure.core/load-one
                  core.clj: 5903  clojure.core/load-one
                  core.clj: 5948  clojure.core/load-lib/fn
                  core.clj: 5947  clojure.core/load-lib
                  core.clj: 5928  clojure.core/load-lib
               RestFn.java:  142  clojure.lang.RestFn/applyTo
                  core.clj:  667  clojure.core/apply
                  core.clj: 5985  clojure.core/load-libs
                  core.clj: 5969  clojure.core/load-libs
               RestFn.java:  137  clojure.lang.RestFn/applyTo
                  core.clj:  667  clojure.core/apply
                  core.clj: 6007  clojure.core/require
                  core.clj: 6007  clojure.core/require
               RestFn.java:  421  clojure.lang.RestFn/invoke
                 state.clj:    1  sausage.state/eval22019/loading--auto--
                 state.clj:    1  sausage.state/eval22019
                 state.clj:    1  sausage.state/eval22019
             Compiler.java: 7176  clojure.lang.Compiler/eval
             Compiler.java: 7165  clojure.lang.Compiler/eval
             Compiler.java: 7635  clojure.lang.Compiler/load
                   RT.java:  381  clojure.lang.RT/loadResourceScript
                   RT.java:  372  clojure.lang.RT/loadResourceScript
                   RT.java:  463  clojure.lang.RT/load
                   RT.java:  428  clojure.lang.RT/load
                  core.clj: 6126  clojure.core/load/fn
                  core.clj: 6125  clojure.core/load
                  core.clj: 6109  clojure.core/load
               RestFn.java:  408  clojure.lang.RestFn/invoke
                  core.clj: 5908  clojure.core/load-one
                  core.clj: 5903  clojure.core/load-one
                  core.clj: 5948  clojure.core/load-lib/fn
                  core.clj: 5947  clojure.core/load-lib
                  core.clj: 5928  clojure.core/load-lib
               RestFn.java:  142  clojure.lang.RestFn/applyTo
                  core.clj:  667  clojure.core/apply
                  core.clj: 5985  clojure.core/load-libs
                  core.clj: 5969  clojure.core/load-libs
               RestFn.java:  137  clojure.lang.RestFn/applyTo
                  core.clj:  667  clojure.core/apply
                  core.clj: 6007  clojure.core/require
                  core.clj: 6007  clojure.core/require
               RestFn.java:  421  clojure.lang.RestFn/invoke
               optical.clj:    1  sausage.optical/eval22013/loading--auto--
               optical.clj:    1  sausage.optical/eval22013
               optical.clj:    1  sausage.optical/eval22013
             Compiler.java: 7176  clojure.lang.Compiler/eval
             Compiler.java: 7165  clojure.lang.Compiler/eval
             Compiler.java: 7635  clojure.lang.Compiler/load
                   RT.java:  381  clojure.lang.RT/loadResourceScript
                   RT.java:  372  clojure.lang.RT/loadResourceScript
                   RT.java:  463  clojure.lang.RT/load
                   RT.java:  428  clojure.lang.RT/load
                  core.clj: 6126  clojure.core/load/fn
                  core.clj: 6125  clojure.core/load
                  core.clj: 6109  clojure.core/load
               RestFn.java:  408  clojure.lang.RestFn/invoke
                  core.clj: 5908  clojure.core/load-one
                  core.clj: 5903  clojure.core/load-one
                  core.clj: 5948  clojure.core/load-lib/fn
                  core.clj: 5947  clojure.core/load-lib
                  core.clj: 5928  clojure.core/load-lib
               RestFn.java:  142  clojure.lang.RestFn/applyTo
                  core.clj:  667  clojure.core/apply
                  core.clj: 5985  clojure.core/load-libs
                  core.clj: 5969  clojure.core/load-libs
               RestFn.java:  137  clojure.lang.RestFn/applyTo
                  core.clj:  667  clojure.core/apply
                  core.clj: 6007  clojure.core/require
                  core.clj: 6007  clojure.core/require
               RestFn.java:  703  clojure.lang.RestFn/invoke
                  core.clj:    1  sausage.core/eval7533/loading--auto--
                  core.clj:    1  sausage.core/eval7533
                  core.clj:    1  sausage.core/eval7533
             Compiler.java: 7176  clojure.lang.Compiler/eval
             Compiler.java: 7165  clojure.lang.Compiler/eval
             Compiler.java: 7635  clojure.lang.Compiler/load
                      REPL:    1  user/eval7529
                      REPL:    1  user/eval7529
             Compiler.java: 7176  clojure.lang.Compiler/eval
             Compiler.java: 7131  clojure.lang.Compiler/eval
                  core.clj: 3214  clojure.core/eval
                  core.clj: 3210  clojure.core/eval
                  main.clj:  414  clojure.main/repl/read-eval-print/fn
                  main.clj:  414  clojure.main/repl/read-eval-print
                  main.clj:  435  clojure.main/repl/fn
                  main.clj:  435  clojure.main/repl
                  main.clj:  345  clojure.main/repl
               RestFn.java: 1523  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   79  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:   55  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:  142  nrepl.middleware.interruptible-eval/interruptible-eval/fn/fn
                  AFn.java:   22  clojure.lang.AFn/run
               session.clj:  171  nrepl.middleware.session/session-exec/main-loop/fn
               session.clj:  170  nrepl.middleware.session/session-exec/main-loop
                  AFn.java:   22  clojure.lang.AFn/run
               Thread.java:  832  java.lang.Thread/run
```
</details>

The relevant part of the stack is 

```
              platform.clj:    6  cljfx.jdk.platform/initialize
              platform.clj:    4  cljfx.jdk.platform/initialize
              platform.clj:   30  cljfx.platform/initialize
              platform.clj:   29  cljfx.platform/initialize
                   api.clj:   75  cljfx.api/eval24166
                   api.clj:   70  cljfx.api/eval24166
```

Which looks like here: https://github.com/cljfx/cljfx/blob/master/src/cljfx/api.clj#L70

``` clojure
(defonce
  ^{:doc "Starts JavaFX runtime and sets implicit exit to false if JavaFX wasn't started
  Either `:cljfx.platform/initialized` or `:cljfx.platform/already-initialized`"}
  initialized
  (platform/initialize))
```

I'm not quire sure what the issue is here.

I then double checked the differences between our projects.. and if I disable AOT compilation, remove `(:gen-class :main true)` and switch to `depstar` then I can generate a jar without crashing (though the jar doesn't run GUI but just drops me into a REPL when I run `java -jar myapp.jar`)

Maybe I'm doing something wrong - but I feel there is some larger issue

**Comment by geokon-gh** [comment]
It seems (according to a conversation I had on IRC) that the issue is that you have a side-effect in a `def`. So this will get evaluated during compilation. Maybe this initialization can happen when the renderer is first launched so that it only happens at runtime?

**Comment by vlaaad** [comment]
Hi!

I guess I'll eventually need to create a FAQ since this issue seems to be brought up from time to time, and the problem is — you can't AOT-compile cljfx without starting JavaFX runtime. I described why it is like that [here](https://github.com/cljfx/cljfx/issues/17#issuecomment-493705534). There, the reported problem is that `lein uberjar` never finishes because cljfx starts javafx application thread that needs to be explicitly terminated. It prevents normal compilation to hang, and it will fail if we compile it on a machine without display.

One workaround that might require changing cljfx was suggested to not start javafx application thread when `*compile-files*` is true (which indicates that compilation is in progress). It still is not a reliable way to AOT-compile cljfx since compilation will fail if you have any class from `javafx.scene.control` package imported in your code. It will help in some cases though, so I might implement it.

I think I'll use a system property like `cljfx.initialize-javafx` so it will not be a breaking change for existing workflows that compile cljfx apps on a machine with a display.

Another possibility that you might want to explore is to compile cljfx with javafx in headless mode, and then use compiled classes with normal javafx dependencies to build an uberjar. It won't be as easy as `lein uberjar`, but if want to give it a try, I can start with how I setup [headless javafx](https://github.com/cljfx/cljfx/blob/master/deps.edn#L26-L27) for tests in github actions.

**Comment by geokon-gh** [comment]
I don't really understand how to integrate the headless javafx into my compilation, sorry. Could you indicate where you are doing that in your [hn](https://github.com/cljfx/hn) project? Because you setup CI there and it seems to work smoothly without any funny stuff.

If it's too involved then I guess I will just skip CI entirely. It'd simplify my workflow but it's not essential. Not something to complicate my life over :)

**Comment by vlaaad** [comment]
I don't do AOT compilation in hn project :)

**Comment by geokon-gh** [comment]
Ah! Right! ...

And I can't get things working without AOT b/c no matter what I try it doesn't seem to be able to find my `-main` (either  `Error: Could not find or load main class sausage.core
` or I get dropped into a REPL). Clojure build tools are a pain in the ass haha :)

I guess I'll stick to just uploading binaries manually. Thanks for the help as always!

**Comment by vlaaad** [comment]
You can use `clojure.main` as a main class that is then calls your main, that's [what hn does](https://github.com/cljfx/hn/blob/master/jpackage/common) — effectively calls `clojure.main -m hn.core`.

**Comment by geokon-gh** [comment]
Ohhhhh, you specify that in the jpackage step. So the intermediary `uberjar` isn't supposed to launch. I should have actually clones the repo and tried it locally :)

**Comment by geokon-gh** [comment]
okay, I got it working.

I get what you meant now. I can run the non-AOT compiled jar with `java -sausage.jar -m sausage.core` - though I'm not sure how I'd begin to bundle that resulting uberjar into something double clickable by a user (though in fairness the AOT uberjars also don't launch on double click)

There were a few gotchas that I'll just mention in case someone else stumbles on this
- `cambada` can't make an uberjar with no AOT compiled main
- `depstars` spits thousands of warnings about duplicate classes b/c of the [crossplatform fix ](https://old.reddit.com/r/Clojure/comments/fw93gc/new_clojurians_ask_anything/fmspx8m/) you gave me on reddit
- so to make the warnings go away I have to make it non-cross-platform again
- because I need to remove the `(:gen-class :main true)` from my `core.clj` file, I can't just have another alias to continue to build AOT locally (other than by uncommenting and then recommenting it before I commit each time)
- the non-AOT app take an eyewatering 17 seconds to launch  vs. 8 seconds with AOT .. which was still bad, but it's not to the point where you're wondering if it crashed or not haha :)

Now that I have a super slow non crossplatform version I guess I need to look into getting `jpackage` working now :)

Sorry about the link-spam. I kept force-pushing the same commit as i tweaked it

**Comment by vlaaad** [comment]
By the way, I don't receive notifications for pushed commits that reference this issue :)

I released 1.7.0 that makes it easier to AOT-compile apps that use cljfx, please give it a try — see [relevant readme section](https://github.com/cljfx/cljfx#aot-compilation-is-complicated).

**Comment by geokon-gh** [comment]
This is fantastic! It works great! I spent all day fighting Github Actions and now I have Releases :) 

Well I'm still trying to figure out how to get a prerelease to track `master` .. but at least I know that the uberjar is getting built and it works :) This will make my life much easier. Thank you so much again!
